dan_layer/engine/tests/state/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// TODO: we should only use stdlib if the template dev needs to include it e.g. use core::mem when stdlib is not
+// available
+
+use common::{generate_abi, TemplateImpl, generate_main};
+use tari_template_abi::{FunctionDef, Type, encode_with_len, decode};
+
+// that's what the example should look like from the user's perspective
+#[allow(dead_code)]
+mod rust {
+    // #[tari::template]
+    pub struct State {
+        value: u32,
+    }
+    
+    // #[tari::impl]
+    impl State {
+        // #[tari::constructor]
+        pub fn new() -> Self {
+            Self { value: 0 }
+        }
+    
+        pub fn set(&mut self, value: u32) {
+            self.value = value;
+        }
+    
+        pub fn get(&self) -> u32 {
+            self.value
+        }
+    }
+}
+
+// TODO: Macro generated code
+#[no_mangle]
+extern "C" fn State_abi() -> *mut u8 {
+    let template_name = "State".to_string();
+
+    let functions = vec![FunctionDef {
+        name: "new".to_string(),
+        arguments: vec![],
+        output: Type::U32, // the component_id
+    }, FunctionDef {
+        name: "set".to_string(),
+        arguments: vec![Type::U32, Type::U32], // the component_id and the new value
+        output: Type::Unit, // does not return anything
+    }, FunctionDef {
+        name: "get".to_string(),
+        arguments: vec![Type::U32], // the component_id
+        output: Type::U32,  // the stored value
+    }];
+
+    generate_abi(template_name, functions)
+}
+
+#[no_mangle]
+extern "C" fn State_main(call_info: *mut u8, call_info_len: usize) -> *mut u8 {
+    let mut template_impl = TemplateImpl::new();
```

Comment:
I like the idea and execution, but the problem with this approach is that it's dynamic, so each time main is called (main is called for every template invocation) a HashMap is allocated. 

This is fine for now though, but I think the macros will generate static code.

Reply:
Agreed. The macros will most likely generate a `match` clause. For now I just wanted a clean way to reuse code.